### PR TITLE
Add ability to compare an attribute value with another

### PR DESCRIPTION
**Pull request by rustamwin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #232

### CODE REVIEW

File: src/Rule/CompareToHandler.php [hunk 1]
```diff
@@ -39,13 +39,17 @@ public function validate(mixed $value, object $rule, ?ValidationContext $context
         }
 
         $result = new Result();
+        $compareAttribute = $rule->getCompareAttribute() ?? $context?->getAttribute() . '_repeat';
```
Comment (xepozz) [inline src/Rule/CompareToHandler.php:42]: What is `_repeat`?

Comment (rustamwin) [reply src/Rule/CompareToHandler.php:42]: Added phpdocs, the logic is ported from Yii2.

Comment (xepozz) [reply src/Rule/CompareToHandler.php:42]: I still don't see a logic of it. If you want to compare `password` and `password_repeat`, then specify values.
```php
class SomeDTO
{
  #[Compare(attribute: 'passwordRepeat')
  private $password;
  #[Compare(attribute: 'password')
  private $passwordRepeat;
}
```

Comment (rustamwin) [reply src/Rule/CompareToHandler.php]: OK. Logic removed.

**Review by xepozz — APPROVED** [review]

File: src/Rule/CompareTo.php [hunk 2]
```diff
@@ -64,9 +64,17 @@ final class CompareTo implements ParametrizedRuleInterface, BeforeValidationInte
...
         /**
-         * @var mixed the constant value to be compared with.
+         * @var mixed the constant value to be compared with. When both this property
+         * and {@see $compareAttribute} are set, this property takes precedence.
          */
-        private $compareValue,
+        private $compareValue = null,
+        /**
+         * @var string|null the name of the attribute to be compared with. When both this property
+         * and {@see $compareValue} are set, the previous one takes precedence.
```
Comment (samdark) [inline src/Rule/CompareTo.php:73]:
```suggestion
         * @var string|null the name of the attribute to be compared with. When both this property
         * and {@see $compareValue} are set, the {@see $compareValue} takes precedence.
```

File: src/Rule/CompareTo.php [hunk 3]
```diff
@@ -130,18 +146,19 @@ public function getOperator(): string
...
+            '<' => 'Value must be less than "{compareValueOrAttribute}".',
+            '<=' => 'Value must be less than or equal to "{compareValueOrAttribute}".',
             default => throw new RuntimeException("Unknown operator: {$this->operator}"),
         };
     }
 
     #[ArrayShape([
-        'compareValue' => '',
+        'compareValue' => 'mixed',
+        'compareAttribute' => '',
```
Comment (samdark) [inline src/Rule/CompareTo.php:161]: Empty?

**Review by samdark — APPROVED** [review]

File: src/Rule/CompareTo.php [hunk 4]
```diff
@@ -152,9 +169,14 @@ public function getOptions(): array
...
         return [
             'compareValue' => $this->compareValue,
+            'compareAttribute' => $this->compareAttribute,
             'message' => [
                 'message' => $this->getMessage(),
-                'parameters' => ['compareValue' => $this->compareValue],
+                'parameters' => [
+                    'compareValue' => $this->compareValue,
+                    'compareAttribute' => $this->compareAttribute,
+                    'compareValueOrAttribute' => $this->compareValue ?? $this->compareAttribute,
```
Comment (vjik) [inline src/Rule/CompareTo.php:178]: Why need this?

File: src/Rule/CompareToHandler.php [hunk 5]
```diff
@@ -39,13 +39,21 @@ public function validate(mixed $value, object $rule, ?ValidationContext $context
...
+        }
+
+        if (!$this->compareValues($rule->getOperator(), $rule->getType(), $value, $compareValue)) {
             $formattedMessage = $this->formatter->format(
                 $rule->getMessage(),
                 [
                     'attribute' => $context?->getAttribute(),
                     'compareValue' => $rule->getCompareValue(),
+                    'compareAttribute' => $rule->getCompareAttribute(),
+                    'compareValueOrAttribute' => $compareValue ?? $compareAttribute,
```
Comment (vjik) [inline src/Rule/CompareToHandler.php:56]: Why need this?

File: src/Rule/CompareTo.php (hunk 4 above)
Comment (rustamwin) [reply src/Rule/CompareTo.php:178]: See rule message.

File: src/Rule/CompareToHandler.php (hunk 5 above)
Comment (rustamwin) [reply src/Rule/CompareToHandler.php:56]: See rule message

File: src/Rule/CompareTo.php (hunk 3 above)
Comment (rustamwin) [reply src/Rule/CompareTo.php]: Fix

**Comment by samdark** [comment]
👍

**Pull request merged**